### PR TITLE
feat(otelgin): enhance gin error tracking with span recording

### PR DESCRIPTION
**Pull request by flc1125**

### CODE REVIEW

**Comment by dmathieu** [comment]
Could you test this change?
By checking there's indeed an error here for example: https://github.com/open-telemetry/opentelemetry-go-contrib/blob/main/instrumentation/github.com/gin-gonic/gin/otelgin/test/gintrace_test.go#L123

Also, I wonder if we need to keep the `gin.errors` attribute. While removing it is a breaking change, keeping it is redundant.

**Comment by dmathieu** [comment]
Side note that otelgin has no owner, so it may be removed soon.
https://github.com/open-telemetry/opentelemetry-go-contrib/issues/6190

**Comment by flc1125** [comment]
> Side note that otelgin has no owner, so it may be removed soon. #6190

Okay, will this package still accept new PRs at present?

**Comment by dmathieu** [comment]
It does.

**Comment by flc1125** [comment]
https://github.com/open-telemetry/opentelemetry-go-contrib/issues/6190

I think I can give it a try.

File: CHANGELOG.md
Comment (dmathieu) [inline CHANGELOG.md:18]:
```suggestion
- Record an error instead of setting the `gin.errors` attribute in `go.opentelemetry.io/contrib/instrumentation/github.com/gin-gonic/gin/otelgin`. (#6346)
```

**Comment by flc1125** [comment]
> Could you test this change? By checking there's indeed an error here for example: [`main`/instrumentation/github.com/gin-gonic/gin/otelgin/test/gintrace_test.go#L123](https://github.com/open-telemetry/opentelemetry-go-contrib/blob/main/instrumentation/github.com/gin-gonic/gin/otelgin/test/gintrace_test.go#L123)
> 
> Also, I wonder if we need to keep the `gin.errors` attribute. While removing it is a breaking change, keeping it is redundant.

It's done.

**Comment by flc1125** [comment]
> Side note that otelgin has no owner, so it may be removed soon.

I spent some time reading through the requirements. I understand that I may still need to make some contributions before applying. If so, I need some time.

File: .golangci.yml
Comment (dmathieu) [inline .golangci.yml]: No, gin instrumentation is not meant to use the SDK.
There is a `test` subfolder which is a separate module. You can put tests that need the SDK in there, they will be ignored by this linter automatically.

**Review by dmathieu — DISMISSED** [review]

File: .golangci.yml
Comment (flc1125) [reply .golangci.yml]: This has been removed before, you can refresh to view the latest code.

File: instrumentation/github.com/gin-gonic/gin/otelgin/gintrace.go
Comment (dmathieu) [inline instrumentation/github.com/gin-gonic/gin/otelgin/gintrace.go:100]:
```suggestion
			span.RecordError(errors.Join(c.Errors))
```

File: instrumentation/github.com/gin-gonic/gin/otelgin/test/gintrace_test.go
Comment (dmathieu) [inline instrumentation/github.com/gin-gonic/gin/otelgin/test/gintrace_test.go:123]: This test already checks for errors. We can check it being recorded here, no need for a new test.
We can also add a check for empty `span.Events` in `TestTrace200`.

**Review by dmathieu — COMMENTED** [review]
This is starting to look good 😸

File: instrumentation/github.com/gin-gonic/gin/otelgin/test/gintrace_test.go
Comment (dmathieu) [inline instrumentation/github.com/gin-gonic/gin/otelgin/test/gintrace_test.go:332]: Shouldn't this test be removed?

Comment (flc1125) [reply instrumentation/github.com/gin-gonic/gin/otelgin/test/gintrace_test.go:332]: Good idea, before this, I hadn't fully read the other tests. This is somewhat redundant.

**Review by dmathieu — APPROVED** [review]

File: instrumentation/github.com/gin-gonic/gin/otelgin/gintrace.go
Comment (pellared) [inline instrumentation/github.com/gin-gonic/gin/otelgin/gintrace.go:100]: Shouldn't we call `span.RecordError` for each error inside `c.Errors`?

Comment (flc1125) [reply instrumentation/github.com/gin-gonic/gin/otelgin/gintrace.go:100]: Can it be like this? I'll try.

Comment (flc1125) [reply instrumentation/github.com/gin-gonic/gin/otelgin/gintrace.go:100]: I tried 2 options to see which one would recommend? My decision is Option 2:

- one:

	```go
	if len(c.Errors) > 0 {
		var errs error
		for _, err := range c.Errors {
			errs = errors.Join(errs, err.Err)
		}
		span.RecordError(errs)
	}
	```

- two:

	```go
	if len(c.Errors) > 0 {
		for _, err := range c.Errors {
			span.RecordError(err.Err)
		}
	}
	```

Comment (dmathieu) [reply instrumentation/github.com/gin-gonic/gin/otelgin/gintrace.go:100]: One would end up being exactly the same thing as we had before. So two.

File: CHANGELOG.md
Comment (pellared) [inline CHANGELOG.md:18]:
```suggestion
- Record errors instead of setting the `gin.errors` attribute in `go.opentelemetry.io/contrib/instrumentation/github.com/gin-gonic/gin/otelgin`. (#6346)
```

File: instrumentation/github.com/gin-gonic/gin/otelgin/gintrace.go
Comment (pellared) [inline instrumentation/github.com/gin-gonic/gin/otelgin/gintrace.go:20]: Can you revert this import rearrange change? We are grouping `go.openteletry.io`. See https://github.com/open-telemetry/opentelemetry-go-contrib/blob/96b1ed1e7b0c1613290df1801ff6ba8474a33ce5/.golangci.yml#L108-L109

File: instrumentation/github.com/gin-gonic/gin/otelgin/test/gintrace_test.go
Comment (pellared) [inline instrumentation/github.com/gin-gonic/gin/otelgin/test/gintrace_test.go:27]: Can you revert this import rearrange change? We are grouping `go.openteletry.io`. See https://github.com/open-telemetry/opentelemetry-go-contrib/blob/96b1ed1e7b0c1613290df1801ff6ba8474a33ce5/.golangci.yml#L108-L109

File: instrumentation/github.com/gin-gonic/gin/otelgin/test/gintrace_test.go
Comment (pellared) [inline instrumentation/github.com/gin-gonic/gin/otelgin/test/gintrace_test.go:126]:
```suggestion
	// verify the error events
```

**Review by pellared — APPROVED** [review]
Just nit comments. Can you please address them before we merge?

File: instrumentation/github.com/gin-gonic/gin/otelgin/test/gintrace_test.go
Comment (flc1125) [reply instrumentation/github.com/gin-gonic/gin/otelgin/test/gintrace_test.go:27]: I'll handle it together later.

**Comment by dmathieu** [comment]
Could you resolve the conflicts?

**Comment by flc1125** [comment]
> Could you resolve the conflicts?

It's done.

File: instrumentation/github.com/gin-gonic/gin/otelgin/gintrace.go
Comment (dmathieu) [inline instrumentation/github.com/gin-gonic/gin/otelgin/gintrace.go:100]: (Sorry, I know this has been approved already)
Should we also set the span status to error?

Comment (flc1125) [reply instrumentation/github.com/gin-gonic/gin/otelgin/gintrace.go:100]: I saw that the setting was here, so I didn't make any settings.

Comment (flc1125) [reply instrumentation/github.com/gin-gonic/gin/otelgin/gintrace.go:100]: Additionally to be supplemented: Even if approved, as long as there is a need for adjustment, it can be pointed out. Try to make this component perfect as much as possible.

Comment (flc1125) [reply instrumentation/github.com/gin-gonic/gin/otelgin/gintrace.go:100]: > However, there are some differences here. In the existing logic, the status is set based on the status code. But in fact, it can also be explained through Error. So there are actually some differences here.

I can try to find some information based on this, look at best practices, see how Status is determined, and then consider how to adjust.

Comment (flc1125) [reply instrumentation/github.com/gin-gonic/gin/otelgin/gintrace.go:100]: https://opentelemetry.io/docs/concepts/signals/traces/#span-status

> To reiterate: Unset represents a span that completed without an error. Ok represents when a developer explicitly marks a span as successful. In most cases, it is not necessary to explicitly mark a span as Ok.

I saw this sentence on the official website, and my understanding is:

As long as we are very sure that there is no problem with him, it can mean OK; as long as there is an error (which may include status codes or may include error), we can set it to ERROR.

Does my understanding make sense?

Comment (dmathieu) [reply instrumentation/github.com/gin-gonic/gin/otelgin/gintrace.go:100]: Could gin have a success HTTP status, but include errors?

Comment (dmathieu) [reply instrumentation/github.com/gin-gonic/gin/otelgin/gintrace.go:100]: Then, we should mark the span as error if the HTTP status is an error, or if otelgin provides errors.

Comment (flc1125) [reply instrumentation/github.com/gin-gonic/gin/otelgin/gintrace.go:100]: Another supplementary material: https://opentelemetry.io/docs/specs/otel/trace/api/#set-status

This is its priority: Ok > Error > Unset

However, I checked the logic of the status codes, and it will only return Error and Unset. Therefore, it will be much easier for me to handle.

Comment (dmathieu) [reply instrumentation/github.com/gin-gonic/gin/otelgin/gintrace.go:100]: Ok is a very specific status that should be set extremely rarely. We can keep Unset.

Comment (flc1125) [reply instrumentation/github.com/gin-gonic/gin/otelgin/gintrace.go:101]: Adjusted

File: instrumentation/github.com/gin-gonic/gin/otelgin/gintrace.go
Comment (dmathieu) [inline instrumentation/github.com/gin-gonic/gin/otelgin/gintrace.go:98]: `gin error` isn't very meaningful.

```suggestion
			span.SetStatus(codes.Error, c.Errors.String())
```

Comment (MrAlias) [reply instrumentation/github.com/gin-gonic/gin/otelgin/gintrace.go:98]: Wont this duplicate, in a less useful flattened string from, the information being added as events below?

File: instrumentation/github.com/gin-gonic/gin/otelgin/test/gintrace_test.go
Comment (MrAlias) [inline instrumentation/github.com/gin-gonic/gin/otelgin/test/gintrace_test.go:201]:
```suggestion
	require.Len(t, events, 2)
```

Avoid the panic.

**Comment by MrAlias** [comment]
> > @dmathieu I could pick up ownership of this module so that it isn't removed, should I just open a PR on the codeowners file?
> 
> Could you review this PR if you want to be an code owner? 😉

cc @akats7 (I think)

**Review by akats7 — APPROVED** [review]

**Comment by flc1125** [comment]
> lgtm, my only real thought is I personally always find it convenient to have the error message at least partially included directly on the span instead of only having the span events. Since span events are sometimes (always?) stored in separate table it could becomes an additional step to view the message depending on the platform you use.
> 
> For example I know a lot of the java instrumentation just adds _error: true_ as an attribute and then creates a span event, and we've had teams complain that the error message has no contextual info

This is indeed a problem, and it usually depends on what backend tools we are using.

Fortunately, when the entire span contains some abnormal events, the tool I am currently using allows me to see the exception information directly in the span interface. Of course, for more detailed information, I still need to switch to view it in Events.

——

Back to this PR, due to the OTEL standard specifications, I suggest we still maintain storing exceptions in Events. After all, this project is for the majority of users, and we need to present ourselves as professionally as possible and in accordance with the standards. 😁

**Comment by akats7** [comment]
We ofc need to record the exception as a span event, what I'm getting at is when it comes to:
```
	span.SetAttributes(attribute.String("gin.errors", c.Errors.String()))

or

	span.SetStatus(codes.Error, "gin error")

```

I do find including the error string to have value

File: instrumentation/github.com/gin-gonic/gin/otelgin/gintrace.go
Comment (flc1125) [reply instrumentation/github.com/gin-gonic/gin/otelgin/gintrace.go:98]: @akats7 We can discuss it here. It should be the same issue here.

Comment (flc1125) [reply instrumentation/github.com/gin-gonic/gin/otelgin/gintrace.go:98]: I recommend using this method to record: `span.SetStatus(codes.Error, c.Errors.String())`

I would like to hear everyone's suggestions.

Comment (pellared) [reply instrumentation/github.com/gin-gonic/gin/otelgin/gintrace.go:98]: From: https://github.com/open-telemetry/opentelemetry-go-contrib/pull/6346#issuecomment-2498987933

> I personally always find it convenient to have the error message at least partially included directly on the span instead of **only** having the span events. Since span events are sometimes (always?) stored in separate table it could becomes an additional step to view the message depending on the platform you use. 
>
> For example I know a lot of the java instrumentation just adds _error: true_ as an attribute and then creates a span event, and we've had teams complain that the error message has no contextual info

Comment (pellared) [reply instrumentation/github.com/gin-gonic/gin/otelgin/gintrace.go:98]: Personally, I do not mind duplication.

Comment (flc1125) [reply instrumentation/github.com/gin-gonic/gin/otelgin/gintrace.go:98]: Based on this reason, if I adjust it like this, is it appropriate?

```go
span.SetStatus(codes.Error, c.Errors.String())
```

File: instrumentation/github.com/gin-gonic/gin/otelgin/test/gintrace_test.go
Comment (pellared) [reply instrumentation/github.com/gin-gonic/gin/otelgin/test/gintrace_test.go]: Reopened 😉

File: instrumentation/github.com/gin-gonic/gin/otelgin/gintrace.go
Comment (pellared) [reply instrumentation/github.com/gin-gonic/gin/otelgin/gintrace.go]: @MrAlias, most of us prefer `span.SetStatus(codes.Error, c.Errors.String())`.

Please create an issue or follow-up PR if you think that this is a wrong decision.

**Pull request merged**
